rfcs/2020-10-15-3480-file-source-rework.md

```diff
@@ -0,0 +1,425 @@
+and corresponding conditional for exiting the source once all files have reached
+EOF. This, along with disabled file discovery, would neatly and simply implement
+the oft-requested batch mode.
+
+#### Observability improvements
+
+There's no grand design here, but we should do the work to go through and
+address all of the relevant issues. Some examples:
+
+* Logging when we see that a file is deleted but need to keep it open
+* Not logging noise around small or empty files
+* Optionally silencing errors due to dangling symlinks
+
+## Doc-level Proposal
+
+* Rename `ignore_older` to `skip_older_than`
+* Replace `start_from_beginning` with `ignore_checkpoints` and `read_from` (see
+    earlier section)
+* Replace `oldest_first` with `mode = tail|tail_sequential|batch`
+* Rename `glob_minimum_cooldown` to `discovery_interval` and disable when `mode = batch`
+
+## Rationale
+
+The file source has dramatically overgrown its original design and is causing
+pain for both users and developers of Vector. It warrants spending some time and
+effort to improve both usability and maintainability. Otherwise, we risk losing
+a large and increasing amount of our time to its maintenance and user support.
+
+By focusing on modularity and simple improvements over a rewrite, we will be
+able to maintain the good parts of the source's long history (accumulated bug
+fixes). This also reduces the risk of introducing new bugs that would be
+inevitable with a more aggressive rewrite.
+
+Modularity also positions us well for the future by encouraging a split between
+user-facing config and implementation-level config. This matches well with our
+config composition RFC and should enable simple config facades for specific
+file-based use cases.
+
+## Drawbacks
+
+This is a backwards incompatible change to the config of one of our most widely
+used sources. To change the interface would likely inconvenience a large number
+of current users. There is also always a risk that bugs are introduced as part
+of refactoring, even though our approach is designed to minimize that risk.
+
+## Alternatives
+
+One alternative would be to rewrite the source from scratch. While this would
+likely result in more maintainable code, it would risk losing much of the
+accumulated knowledge present in the existing implementation. It would also
+likely take much longer and present a more difficult transition plan than the
+evolution proposed here.
+
+Another alternative would be to leave the implementation largely alone and focus
+on improving our documentation and config UI alone. While this would likely
+yield strong benefits, it would leave a number of important issues unaddressed
+and do little for our ability to extend the source in the future.
+
+## Outstanding Questions
+
+- Should we change the user-facing config at the same time at the implementation
+    or split the two?
+- How to handle transitioning users to the new interface?
```

Comment:
We have some precedence for this in the `lua` transform with the `version` option. This is a required option going forward and the user must set it to `2`.